lockbox-ios/Store/AccountStore.swift

```diff
@@ -0,0 +1,139 @@
+         keychainWrapper: KeychainWrapper = KeychainWrapper.standard) {
+        self.dispatcher = dispatcher
+        self.keychainWrapper = keychainWrapper
+
+        self.dispatcher.register
+                .filterByType(class: AccountAction.self)
+                .subscribe(onNext: { action in
+                    switch action {
+                    case .oauthRedirect(let url):
+                        self.oauthLogin(url)
+                    case .clear:
+                        self.clear()
+                    }
+                })
+                .disposed(by: self.disposeBag)
+
+        if let accountJSON = self.keychainWrapper.string(forKey: KeychainKey.accountJSON.rawValue) {
+            self.fxa = try? FirefoxAccount.fromJSON(state: accountJSON)
+            self.generateInitialURL()
+            self.populateInitialValues()
+        } else {
+            FxAConfig.release { (config: FxAConfig?, _) in
+                if let config = config {
+                   self.fxa = try? FirefoxAccount(
+                           config: config,
+                           clientId: Constant.fxa.clientID,
+                           redirectUri: Constant.fxa.redirectURI)
+
+                    self.generateInitialURL()
+                    self.populateInitialValues()
+                }
+            }
+        }
+    }
+}
+
+extension AccountStore {
+    private func generateInitialURL() {
+        self.fxa?.beginOAuthFlow(scopes: Constant.fxa.scopes, wantsKeys: true) { url, _ in
+            if let url = url {
+                self._loginURL.onNext(url)
+            }
+        }
+    }
+
+    private func populateInitialValues() {
+        self.fxa?.getOAuthToken(scopes: Constant.fxa.scopes) { (info: OAuthInfo?, _) in
+            self._oauthInfo.onNext(info)
+        }
+
+        self.fxa?.getProfile { (profile: Profile?, _) in
+            self._profile.onNext(profile)
+        }
+    }
+
+    private func clear() {
+        for identifier in KeychainKey.allValues {
+            _ = self.keychainWrapper.removeObject(forKey: identifier.rawValue)
+        }
+
+        self._profile.onNext(nil)
+        self._oauthInfo.onNext(nil)
+    }
+
+    private func oauthLogin(_ navigationURL: URL) {
+        guard let components = URLComponents(url: navigationURL, resolvingAgainstBaseURL: true),
+              let queryItems = components.queryItems else {
+            return
+        }
+
+        var dic = [String: String]()
+        queryItems.forEach {
+            dic[$0.name] = $0.value
+        }
+
+        guard let code = dic["code"],
+              let state = dic["state"] else {
+            return
+        }
+
+        self.fxa?.completeOAuthFlow(code: code, state: state) { (info: OAuthInfo?, _) in
+            self._oauthInfo.onNext(info)
+
+            if let accountJSON = try? self.fxa?.toJSON(), accountJSON != nil {
+                self.keychainWrapper.set(accountJSON!, forKey: KeychainKey.accountJSON.rawValue)
+            }
```

Comment:
Nit: everytime `!` is used in swift, an angel loses its wings.

Also: `??` should collapse under its own uncertainty.

```swift
            if let opt = try? self.fxa?.toJSON(), let accountJSON = opt {
                self.keychainWrapper.set(accountJSON, forKey: KeychainKey.accountJSON.rawValue)
            }
```